micro_sam/sam_annotator/_widgets.py

```diff
@@ -0,0 +1,58 @@
+from enum import Enum
+from pathlib import Path
+from typing import TYPE_CHECKING, Optional
+
+from magicgui import magic_factory
+from magicgui.tqdm import tqdm
+from superqt.utils import thread_worker
+
+from micro_sam.util import (
+    ImageEmbeddings,
+    get_sam_model,
+    precompute_image_embeddings,
+    _MODEL_URLS,
+    _DEFAULT_MODEL,
+    _available_devices,
+)
+
+if TYPE_CHECKING:
+    import napari
+
+Model = Enum("Model", _MODEL_URLS)
+available_devices_list = ["auto"] + _available_devices()
+
+
+@magic_factory(call_button="Compute image embeddings",
+               device = {"choices": available_devices_list})
+def embedding_widget(
+    image: "napari.layers.Image",
+    model: Model = Model.__getitem__(_DEFAULT_MODEL),
+    device = "auto",
+    save_path: Optional[Path] = None,  # where embeddings for this image are cached (optional)
+    custom_model: Optional[str] = None,  # A filepath or URL to custom model weights.
+) -> ImageEmbeddings:
+    """Image embedding widget."""
+    # for access to the predictor and the image embeddings in the widgets
+    global PREDICTOR, IMAGE_EMBEDDINGS
+    # Initialize the model
+    PREDICTOR = get_sam_model(model_type=model.name)
+
+    # Get image dimensions
+    if not image.rgb:
+        ndim = image.data.ndim
+    else:
+        # assumes RGB channels are the last dimension
+        ndim = image.data.ndim
+
+    # Compute the embeddings for the image data
+    with tqdm() as pbar:
+        @thread_worker(connect={"finished": lambda: pbar.progressbar.hide()})
+        def _compute_image_embedding(PREDICTOR, image, save_path, ndim=None):
+            IMAGE_EMBEDDINGS = precompute_image_embeddings(
+                predictor = PREDICTOR,
+                input_ = image,
+                save_path = str(save_path),
+                ndim=ndim,
+            )
+
+        _compute_image_embedding(PREDICTOR, image.data, save_path, ndim=ndim)
```

Comment:
Something else to check - does this produce sensible results if I pass it an RGB image data array, and ndim = image.data.ndim - 1? I don't need to do anything fancy to the actual image data array, do I?

Reply:
When I give it an RGB image, like the astronaut example image, it seems to take a really long time (I've been closing the window after about a minute, maybe it never completes?)

In contrast, a grayscale image of the same size, like the camera example image, has the embedding computation finished in a few seconds.

Reply:
Yes, for RGB data it should be enough to set `ndim = image.data.ndim - 1`. See also the comment below. Once you fix that it should compute the embeddings for RGB images in a few seconds as well.
(Currently it interprets it as a 3d volume, and that's why it takes so long.)